stan/math/rev/functor/adj_jac_apply.hpp

```diff
@@ -547,54 +464,70 @@ struct adj_jac_vari : public vari {
- * arguments indicating whether or not the adjoint^T Jacobian product must be
- * computed for each input argument. This argument is passed to operator() so
- * that any unnecessary preparatory calculations for multiply_adjoint_jacobian
- * can be avoided if possible.
+ * adj_jac_apply takes as a template argument a functor F of the form
+ * ```
+ * template <typename T1, typename T2> // Any number of templates
+ * class my_functor {
+ *  // Helper type trait that is an eigen map for containers or scalar
+ *  // for scalars
+ *  adj_arg_t<T1> A_;
+ *  adj_arg_t<T2> B_;
+ *  // etc.
  *
- * (required) std::tuple<T1, T2, ...> multiply_adjoint_jacobian(const
- * std::array<bool, size>& needs_adj, const Eigen::VectorXd& adj)
+ *  // Constructor to allocate memory
+ *  // Must accept same types as T1 and T2
+ *  my_functor(const T1& A, const T2& B) :
+ *    A_(setup_adj_arg<T1>(A.rows(), A.cols()))
+ *    B_(setup_adj_arg<T2>(B.rows(), B.cols())) {}
+ *  // Member function to calculate forward pass values
+ *  // Must accept same types as the result of calling value_of() on
+ *  //  T1 and T2 types
+ *  template <typename S1, typename S2>
+ *  auto forward_pass(S1&& A, S2&& B) {
+ *    // Calculate forward pass value
+ *  }
+ *  // Calculate adjoint values in reverse pass
+ *  // Argument type must be the same as result of forward_pass() and
+ *  // is the result of dL/dy
+ *  template <typename S1>
+ *  auto reverse_pass(S1&& adj) {
+ *  // Calculate adjoints
+ *  // Must return a tuple of the adjoints for the inputs
+ *  return std::make_tuple(adjoint_a, adjoint_b);
+ *  // If one adjoint does not need calculated return back a nullptr
+ *  // in it's position such as
+ *  return std::make_tuple(adjoint_a, nullptr); // Don't compute adjoint of B
+ *  }
+ * };
+ * ```
+ * The template parameters can be any type besides containers of containers.
  *
- * where T1, T2, etc. are the same types as in operator(), needs_adj is the
- * same as in operator(), and adj is the vector dL/dy.
- *
- * operator() is responsible for computing f(x) and multiply_adjoint_jacobian
+ * `forward_pass()`` is responsible for computing f(x) and `reverse_pass()`
  * is responsible for computing the necessary adjoint transpose Jacobian
  * products (which frequently does not require the calculation of the full
  * Jacobian).
  *
- * operator() will be called before multiply_adjoint_jacobian is called, and
+ * `forward_pass()`` will be called before `reverse_pass()` is called, and
  * is only called once in the lifetime of the functor
- * multiply_adjoint_jacobian is called after operator() and may be called
+ * `reverse_pass()` is called after `forward_pass()` and may be called
  * multiple times for any single functor
  *
  * The functor supplied to adj_jac_apply must be careful to allocate any
  * variables it defines in the autodiff arena because its destructor will
  * never be called and memory will leak if allocated anywhere else.
```

Comment:
Maybe mention `adj_arg_t` is recommended to make this allocation easy, though it can be done manually.